File: dart/lib/src/tokens/token_formatter_builder.dart
import 'package:cucumber_gherkin/src/language/token.dart';
import 'package:cucumber_gherkin/src/parser/builder.dart';
import 'package:cucumber_gherkin/src/parser/rule_type.dart';
import 'package:cucumber_gherkin/src/tokens/no_op_result.dart';
import 'package:cucumber_gherkin/src/tokens/test_token_formatter.dart';

/// A [Builder] that renders each scanned token using the shared `*.tokens`
/// text format used by the acceptance tests.
class TokenFormatterBuilder implements Builder<NoOpResult> {
  /// The formatter used to render individual tokens.
  final TestTokenFormatter formatter = TestTokenFormatter();

  /// Accumulates the rendered token lines.
  final StringBuffer tokensTextBuilder = StringBuffer();

  /// The full rendered token text accumulated so far.
  String get tokensText => tokensTextBuilder.toString();

  @override
  void build(Token token) {
    tokensTextBuilder.writeln(formatter.formatToken(token));
  }

  @override
  void endRule(RuleType ruleType) {
    // nop
  }

  @override
  void reset() {
    // nop
  }

  @override
  NoOpResult get result => NoOpResult();

  @override
  void startRule(RuleType ruleType) {
    // nop
  }
}

File: java/src/test/java/io/cucumber/gherkin/TokenFormatterBuilder.java
package io.cucumber.gherkin;

import java.util.ArrayList;
import java.util.List;

class TokenFormatterBuilder implements Parser.Builder<List<String>> {
    private final TokenFormatter formatter = new TokenFormatter();
    private final List<String> tokens = new ArrayList<>();

    @Override
    public void build(Token token) {
        tokens.add(formatter.formatToken(token));
    }

    @Override
    public void startRule(Parser.RuleType ruleType) {
    }

    @Override
    public void endRule(Parser.RuleType ruleType) {
    }

    @Override
    public List<String> getResult() {
        return tokens;
    }

    @Override
    public void reset(String uri) {
    }
}
